### PR TITLE
Fix #186: Crypto 3.0: Custom activation

### LINKED ISSUE #186: Crypto 3.0: Custom activation
Implement support for custom ECIES based activation as defined in specification:

### Layer 1

- Encrypted by ECIES at app scope
- Layer can be decrypted at the REST application server.
    ```json
    {
        "requestObject": {
            "type": "custom",
            "identityAttributes": {
                "login" : "...",
                "password" : "...",
                "arbitrary-key" : "value-dictionary"
            },
            "activationData": {
                "ephemeralKey"  : "BASE64",
                "encryptedData" : "BASE64",
                "mac"           : "BASE64"
            }
        }
    }
    ```
- PowerAuth server will return encrypted response:
  ```json
  {
      "encryptedData" : "BASE64",
      "mac"           : "BASE64"
  }
  ```
  Note that the response must be once more time encrypted with key derived for Layer 1.

### `CreateActivation` (custom activation)

- `nonce` must be validated to be non-empty

```xml
<xs:element name="CreateActivationRequest">
    <xs:annotation>
        <xs:documentation>Request for creating a new activation directly.</xs:documentation>
    </xs:annotation>
    <xs:complexType>
        <xs:sequence>
            <xs:element name="userId" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="timestampActivationExpire" type="xs:dateTime" minOccurs="0" maxOccurs="1"/>
            <xs:element name="maxFailureCount" type="xs:long" minOccurs="0" maxOccurs="1"/>
            <!-- the rest is same as for regular activation -->
            <xs:element name="applicationKey" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="ephemeralKey" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="encryptedData" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="mac" type="xs:string" minOccurs="1" maxOccurs="1"/>
        </xs:sequence>
    </xs:complexType>
</xs:element>

<xs:element name="CreateActivationResponse">
    <xs:annotation>
        <xs:documentation>Response for creating a new activation directly.</xs:documentation>
    </xs:annotation>
    <xs:complexType>
        <xs:sequence>
            <xs:element name="activationId" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="encryptedData" type="xs:string" minOccurs="1" maxOccurs="1"/>
            <xs:element name="mac" type="xs:string" minOccurs="1" maxOccurs="1"/>
        </xs:sequence>
    </xs:complexType>
</xs:element>
```

### PR DESCRIPTION
**Pull request by romanstrobl**
This PR introduces custom activation implementation on PowerAuth server.

I added the timestamp expiration check which was previously missing to v2 version, too.

### CODE REVIEW

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 1]
```diff
@@ -205,7 +205,34 @@ public PrepareActivationResponse prepareActivation(PrepareActivationRequest requ
     @Override
     @Transactional
     public CreateActivationResponse createActivation(CreateActivationRequest request) throws GenericServiceException {
-        throw new IllegalStateException("Not implemented yet.");
+        try {
+            // Get request parameters
+            String userId = request.getUserId();
+            Date activationExpireTimestamp = XMLGregorianCalendarConverter.convertTo(request.getTimestampActivationExpire());
```
Comment (petrdvorak) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: `request.getTimestampActivationExpire()` could be `null`.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 2]
```diff
@@ -205,7 +205,34 @@ public PrepareActivationResponse prepareActivation(PrepareActivationRequest requ
...
+            Long maxFailedCount = request.getMaxFailureCount();
```
Comment (petrdvorak) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: Could I double check where the default value (`5`) is set? In general, I believe we should do some basic request validation here, to make the future work on wultra/powerauth-server#9 a bit easier.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java [hunk 3]
```diff
@@ -205,7 +205,34 @@ public PrepareActivationResponse prepareActivation(PrepareActivationRequest requ
...
+            String applicationKey = request.getApplicationKey();
+            byte[] ephemeralPublicKey = BaseEncoding.base64().decode(request.getEphemeralPublicKey());
+            byte[] mac = BaseEncoding.base64().decode(request.getMac());
```
Comment (petrdvorak) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: What if the request values are `null`?

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java (hunk 1 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: Yes, this is handled as previously in the `initActivation()` method. By default the value is `null` and it is non-null only when it is modified.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java (hunk 2 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: Yes, this is handled as previously in the `initActivation()` method. By default the value is `null` and it is non-null only when it is modified.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java (hunk 3 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: Good catch, I will fix it in all occurrences.

Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: I will add new error `INVALID_REQUEST` and add null checks to all methods in `PowerAuthServiceImpl`, such validation is generally missing.

Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/v3/PowerAuthServiceImpl.java]: I added nullability checks into all service methods, added the error as well as fixed missing error messages for other errors.

**Pull request merged**